Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+{
+    public ValueTask<ShoppingCartModel?> Find(Guid id, CancellationToken ct) =>
+        dbContext.FindAsync<ShoppingCartModel>([id], ct);
+
+    public Task Store(Guid id, ShoppingCartModel? current, ShoppingCartEvent[] events, CancellationToken ct)
+    {
+        foreach (var @event in events)
+        {
+            switch (@event, current)
+            {
+                case (Opened(var clientId, var openedAt), null):
+                    dbContext.Add(new ShoppingCartModel
+                    {
+                        Id = id,
+                        ClientId = clientId,
+                        Status = ShoppingCartModel.ShoppingCartStatus.Pending,
+                        OpenedAt = openedAt
+                    });
+                    break;
+                case (ProductAdded(var (productId, quantity, price), var addedAt), not null):
+                    var toUpdate = current.ProductItems.SingleOrDefault(p => p.ProductId == productId);
+
+                    if (toUpdate == null)
+                    {
+                        current.ProductItems.Add(new ShoppingCartModel.PricedProductItem
+                        {
+                            ProductId = productId, Quantity = quantity, Price = price, AddedAt = addedAt,
+                        });
+                    }
+                    else
+                    {
+                        toUpdate.Quantity += quantity;
+                        toUpdate.UpdatedAt = addedAt;
+                    }
+
+                    break;
+                case (ProductRemoved(var (productId, quantity, price), var removedAt), not null):
+                    var existing = current.ProductItems.Single(p => p.ProductId == productId);
+
+                    if (existing.Quantity - quantity == 0)
+                    {
+                        current.ProductItems.Remove(existing);
+                    }
+                    else
+                    {
+                        existing.Quantity -= quantity;
+                        existing.UpdatedAt = removedAt;
+                    }
+
+                    break;
+                case (Confirmed (var clientId, var confirmedAt), not null):
+                    current.Status = ShoppingCartModel.ShoppingCartStatus.Confirmed;
+                    current.ClientId = clientId;
+                    current.ConfirmedAt = confirmedAt;
+                    dbContext.ShoppingCarts.Update(current);
+                    break;
+                case (Cancelled, not null):
+                    dbContext.ShoppingCarts.Remove(current);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException(nameof(@event));
+            }
+        }
+
+        return dbContext.SaveChangesAsync(ct);
+    }
+}
+
+public interface IProductPriceCalculator
+{
+    IReadOnlyList<PricedProductItem> Calculate(params ProductItem[] productItems);
```

Comment:
no params required (the less frilly features touched the better, no?)

Reply:
Fine, especially with new collection expressions. Updated.